### PR TITLE
Add options to leak and cleanup resources

### PR DESCRIPTION
**Pull request by TheSpiritXIII**
Provides the ability to leak and cleanup resources for an improved debugging experience, especially when running tests locally.

### CODE REVIEW

**Comment by bernot-dev** [comment]
The pattern I typically see is to "leak" by default, and then cleanup by default before the test starts. This can prevent the case where someone leaks resources that impact subsequent tests.

**Comment by TheSpiritXIII** [comment]
Hi @bernot-dev can you share some examples where you see this convention?

My reasoning for not leaking by default are:

- It doesn't change behavior of tests today.
- It doesn't mess up your cluster if you run it locally by default.
- It makes tests significantly slower. When we clean after a test, we know what to clean (at least after we merge #602). If we clean before each test run, we don't know what to clean. This translates to hundreds of blocking API calls. We have to lookup all Kubernetes API groups, and then individually clean each group. Note, we cannot parallelize these because the Kubernetes API would only throttle us.

**Comment by bernot-dev** [comment]
> * It doesn't mess up your cluster if you run it locally by default.

I guess it's still a little strange to me that E2E tests wouldn't be running in a cluster dedicated to testing. But yes, it does speed things up. Probably a good tradeoff for us.

> * Tests would fail since they expect resources from the prior one to get cleaned up. The `leak` flag would only be useful for subtests right now.

True, if the cleanup didn't happen before running tests.

> * It makes tests significantly slower. When we clean after a test, we know what to clean (at least after we merge [Add Kubernetes clients which tracks objects created in e2e tests #602](https://github.com/GoogleCloudPlatform/prometheus-engine/pull/602)). If we clean before each test run, we don't know what to clean. This translates to hundreds of blocking API calls. We have to lookup all Kubernetes API groups, and then individually clean each group. Note, we cannot parallelize these because the Kubernetes API would only throttle us.

Makes sense. I would think hundreds of queries would be on the order of seconds, though, as long as they don't have to deal with finalizers and such.

**Comment by TheSpiritXIII** [comment]
@bernot-dev please take another look! I flipped the defaults of the flags, as requested. I got around the performance issue by simply defaulting turning the cleanup flag off when running under Docker (which implicitly does cleanup by recreating a new `kind` cluster).

File: e2e/operator_context_test.go [hunk 1]
```diff
@@ -66,7 +67,7 @@ const (
 
 var (
 	startTime    = time.Now().UTC()
-	globalLogger = zap.New(zap.Level(zapcore.Level(-1)))
+	globalLogger = zap.New(zap.Level(zapcore.DebugLevel))
```
Comment (bernot-dev) [inline e2e/operator_context_test.go:70]: 👍

**Review by bernot-dev — APPROVED** [review]
LGTM

**Pull request merged**